contracts/generator_controller/src/contract.rs

```diff
@@ -405,6 +423,29 @@ fn tune_pools(deps: DepsMut, env: Env) -> ExecuteResult {
         config.pools_limit,
     )?;
 
+    // Set allocation points for the main pool
+    if let Some(main_pool) = config.main_pool {
+        // Main pool may appear in the pool list thus we need to eliminate its contribution in the total VP
+        tune_info
+            .pool_alloc_points
+            .retain(|(pool, _)| pool != &main_pool.to_string());
+
+        let total_vp: Uint128 = tune_info
+            .pool_alloc_points
+            .iter()
+            .fold(Uint128::zero(), |acc, (_, vp)| acc + vp);
+        // Calculate main pool contribution.
+        // Example (30% for the main pool): VP + x = y, x = 0.3y => y = VP/0.7  => x = 0.3 * VP / 0.7,
+        // where VP - total VP, x - main pool's contribution, y - new total VP.
+        // x = 0.3 * VP * (1-0.3)^(-1)
+        let main_pool_contribution = config.main_pool_min_alloc
+            * total_vp
+            * (Decimal::one() - config.main_pool_min_alloc).inv().unwrap();
```

Comment:
I guess, you hope that there will never be zero and therefore use unwrap function. Maybe it is better to use the operator **?** and handle the error?

Reply:
Yes, we guarantee this in this [line](https://github.com/astroport-fi/astroport-governance/blob/67cb307546643ca0c890d05252a92a56699a8125/contracts/generator_controller/src/contract.rs#L509)